### PR TITLE
Initial Code Dump

### PR DESCRIPTION
**Pull request by marstr**
This begins an effort to refresh the Android SDK to reflect the existing App Center Push SDK. This is being done for multiple reasons:
1. We've gotten a lot of positive feedback about the App Center Push SDKs, especially in regards to how they stack up against the existing offerings.
1. It has been announced that App Center Push is being retired. Before it is officially done, we'd like there to be a dead-simple migration for App Center Push users.

This etches out a skeleton with some of the public-facing surface area that we expect to have in the next iteration of this SDK.

### CODE REVIEW

File: notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/BagMiddleware.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Does no routing of Middleware based on conditions, simply applies all enrichments it is aware of
+ * then calls the next Middleware.
+ */
+class BagMiddleware implements InstallationMiddleware {
+
+    private final Set<InstallationEnricher> enrichers;
```
Comment (mpodwysocki) [inline notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/BagMiddleware.java:13]: Let's be consistent with our member variables, prepending `m` for example `mEnrichers`

File: notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/Installation.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package com.microsoft.windowsazure.messaging.notificationhubs;
+
+/**
+ * Wraps a record of a device, for use resolving future requests to broadcast notifications.
+ */
+public class Installation {
+    private String pushChannel;
```
Comment (mpodwysocki) [inline notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/Installation.java:7]: Let's be consistent with our member variables, prepending `m` for example `mPushChannel`

File: notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/InstallationManager.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.microsoft.windowsazure.messaging.notificationhubs;
+
+import com.microsoft.windowsazure.messaging.notificationhubs.async.NotificationHubFuture;
+
+public interface InstallationManager {
+    NotificationHubFuture<String> saveInstallation(Installation installation);
```
Comment (mpodwysocki) [inline notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/InstallationManager.java:6]: What is the string being returned here and how is it useful? Some documentation should be needed here

File: notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/NotificationHub.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+ * A Singleton controller that wraps all interactions with Firebase Cloud Messaging and Azure
+ * Notification Hubs.
+ */
+public final class NotificationHub {
+    private static NotificationHub sInstance;
+
+    private NotificationListener mListener;
+    private final List<InstallationMiddleware> mMiddleware;
+    private final PushChannelEnricher pushChannelEnricher;
+    private InstallationManager manager;
```
Comment (mpodwysocki) [inline notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/NotificationHub.java:22]: Mix of `m` and non `m` prefixed variables.  Let's stick with prepending `m`

File: notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/PushChannelEnricher.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package com.microsoft.windowsazure.messaging.notificationhubs;
+
+public class PushChannelEnricher implements InstallationEnricher {
+    private String channel;
```
Comment (mpodwysocki) [inline notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/PushChannelEnricher.java:4]: Consistency in naming for member variables

File: notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/DefaultMiddleware.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package com.microsoft.windowsazure.messaging.notificationhubs;
+
+/**
+ * Wraps an {@link InstallationEnricher} to fulfill the most basic responsibilities of middleware.
+ */
+class DefaultMiddleware implements InstallationMiddleware {
+
+    private InstallationEnricher enricher;
```
[the remark this answers is not in the document]
Comment (marstr) [reply notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/DefaultMiddleware.java:8]: Fixed in 583cdf30d92d7f9a3962e8b37fae21d9037b4511

File: notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/Installation.java (hunk 2 above)
Comment (marstr) [reply notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/Installation.java:7]: Fixed in 583cdf30d92d7f9a3962e8b37fae21d9037b4511

File: notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/NotificationHub.java (hunk 4 above)
Comment (marstr) [reply notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/NotificationHub.java:22]: Fixed in 583cdf30d92d7f9a3962e8b37fae21d9037b4511

File: notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/PushChannelEnricher.java (hunk 5 above)
Comment (marstr) [reply notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/PushChannelEnricher.java:4]: Fixed in 583cdf30d92d7f9a3962e8b37fae21d9037b4511

File: notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/InstallationManager.java (hunk 3 above)
Comment (marstr) [reply notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/InstallationManager.java:6]: Fixed in b5039971c5e48447912617bb42561b6e3cdc88c1

**Pull request merged**